### PR TITLE
Split source files, store `data` component as a `Matrix`, instead of a `Vector`

### PR DESCRIPTION
**Pull request by dmbates**
- split code for `cholesky.jl`, `hermitian.jl`, and `triangular.jl` from `RectangularFullPacked.jl`
- declare the `data` field as `Matrix{T}` instead of `Vector{T}`.
    * doing so makes it easier to determine dimensions and also allows for examining the `parent` array in a natural way
- create abstract type `AbstractRFP` and utilities for determining dimensions
    * determining dimensions and converting index pairs in the original array to pairs in the parent array can be written for the abstract type
- Keep track of conjugation of complex values in the packed matrices.
- Allow for `transr` to be `T` (`C` for complex types).

### CODE REVIEW

**Comment by dmbates** [comment]
@andreasnoack This is still a draft so there is no need to review in detail.  However, could you tell me if it look like this is headed in the right direction?

One question I would have is whether it is worthwhile typing the `data` field as `StridedMatrix{T}`.  I can't picture a situation when it would be other than a `Matrix{T}`.  Would it seem okay to use the more restricted type or would that come back to haunt us?

Today I will do some more digging on the transposed blocks being stored as the conjugate transpose.

**Comment by andreasnoack** [comment]
> One question I would have is whether it is worthwhile typing the data field as StridedMatrix{T}

I don't think it should be `StridedMatrix{T}` since it's a non-concrete type so we'd introduce dynamic dispatch for anything using these structs. In theory, we could make the array type a type parameter but I doubt it's worth it. I think restricting it to `Matrix{T}` is the better solution.

**Comment by dmbates** [comment]
> > One question I would have is whether it is worthwhile typing the data field as StridedMatrix{T}
> 
> I don't think it should be `StridedMatrix{T}` since it's a non-concrete type so we'd introduce dynamic dispatch for anything using these structs. In theory, we could make the array type a type parameter but I doubt it's worth it. I think restricting it to `Matrix{T}` is the better solution.

Thanks, I will make that change.

**Comment by dmbates** [comment]
BTW, if you are wondering why the branch is called `mtating` instead of `mutating`, it's a typo.

**Comment by andreasnoack** [comment]
Is this ready for review?

**Comment by dmbates** [comment]
> Is this ready for review?

You could review it now if you wish.  It does cover the cases that were previously allowed.  I think that I can incorporate the `transa` option if I spend a day or two experimenting with when conjugate transposes occur.

Can you wait for me to do that?

**Comment by dmbates** [comment]
@andreasnoack, @arslan I think this is ready to review

**Comment by ararslan** [comment]
I'm flattered that you think I'm in any way qualified to review this. 😅 I can try to work my way through it but it will take a while. Andreas will surely be able to provide a more meaningful review.

File: src/cholesky.jl [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+(\)(A::CholeskyRFP, B::StridedVecOrMat) =
+    LAPACK_RFP.pftrs!(A.transr, A.uplo, A.data, copy(B))
+(\)(A::HermitianRFP, B::StridedVecOrMat) = cholesky(A) \ B
+
+LinearAlgebra.inv!(A::CholeskyRFP) =
+    HermitianRFP(LAPACK_RFP.pftri!(A.transr, A.uplo, A.data), A.transr, A.uplo)
+LinearAlgebra.inv(A::CholeskyRFP) = LinearAlgebra.inv!(copy(A))
+LinearAlgebra.inv(A::HermitianRFP) = LinearAlgebra.inv!(cholesky(A))
+
+LinearAlgebra.parent(A::CholeskyRFP) = A.data
```
Comment (andreasnoack) [inline src/cholesky.jl]: Maybe let's drop this definition. If you need to work with the field then you might as well just access it directly.

File: src/hermitian.jl [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    rs, doconj = _packedinds(A, Int(i), Int(j), iseven(n), l)
+    val = A.data[first(rs), last(rs)]
+    return doconj ?  conj(val) : val
+end
+
+function Ac_mul_A_RFP(A::Matrix{T}, uplo = :U) where {T<:BlasFloat}
+    n = size(A, 2)
+    par = Matrix{T}(undef, _parentsize(n))
+    tr = T <: Complex ? 'C' : 'T'
+    (ul = first(string(uplo))) ∈ "UL" || throw(ArgumentError("uplo must be either :L or :U"))
```
Comment (andreasnoack) [inline src/hermitian.jl]:
```suggestion
    ul = first(string(uplo))
    if ul ∉ "UL"
        throw(ArgumentError("uplo must be either :L or :U"))
    end
```
For readability and better coverage reporting

File: src/lapack.jl [hunk 3]
```diff
@@ -26,7 +26,7 @@ for (f, elty, relty) in (
             alpha::Real,
             A::StridedMatrix{$elty},
             beta::Real,
-            C::StridedVector{$elty},
+            C::StridedVecOrMat{$elty},
```
Comment (andreasnoack) [inline src/lapack.jl]:
```suggestion
            C::VecOrMat{$elty},
```
This is probably too wide a signature. The routine requires that all the memory is contiguous so I think this should be narrowed to VecOrMat

File: src/lapack.jl [hunk 4]
```diff
@@ -77,7 +77,7 @@ for (f, elty) in (
 )
 
     @eval begin
-        function pftrf!(transr::Char, uplo::Char, A::StridedVector{$elty})
+        function pftrf!(transr::Char, uplo::Char, A::StridedVecOrMat{$elty})
```
Comment (andreasnoack) [inline src/lapack.jl]:
```suggestion
        function pftrf!(transr::Char, uplo::Char, A::VecOrMat{$elty})
```

File: src/lapack.jl [hunk 5]
```diff
@@ -106,7 +106,7 @@ for (f, elty) in (
 )
 
     @eval begin
-        function pftri!(transr::Char, uplo::Char, A::StridedVector{$elty})
+        function pftri!(transr::Char, uplo::Char, A::StridedVecOrMat{$elty})
```
Comment (andreasnoack) [inline src/lapack.jl]:
```suggestion
        function pftri!(transr::Char, uplo::Char, A::VecOrMat{$elty})
```

File: src/lapack.jl [hunk 6]
```diff
@@ -139,7 +139,7 @@ for (f, elty) in (
         function pftrs!(
             transr::Char,
             uplo::Char,
-            A::StridedVector{$elty},
+            A::StridedVecOrMat{$elty},
```
Comment (andreasnoack) [inline src/lapack.jl]:
```suggestion
            A::VecOrMat{$elty},
```

File: src/lapack.jl [hunk 7]
```diff
@@ -196,18 +196,20 @@ for (f, elty) in (
             trans::Char,
             diag::Char,
             alpha::$elty,
-            A::StridedVector{$elty},
+            A::StridedMatrix{$elty},
```
Comment (andreasnoack) [inline src/lapack.jl]:
```suggestion
            A::Matrix{$elty},
```

File: src/lapack.jl [hunk 8]
```diff
@@ -256,10 +258,12 @@ for (f, elty) in (
 )
 
     @eval begin
-        function tftri!(transr::Char, uplo::Char, diag::Char, A::StridedVector{$elty})
+        function tftri!(transr::Char, uplo::Char, diag::Char, A::StridedMatrix{$elty})
```
Comment (andreasnoack) [inline src/lapack.jl]:
```suggestion
        function tftri!(transr::Char, uplo::Char, diag::Char, A::Matrix{$elty})
```

File: src/lapack.jl [hunk 9]
```diff
@@ -295,11 +299,15 @@ for (f, elty) in (
 )
 
     @eval begin
-        function tfttr!(transr::Char, uplo::Char, Arf::StridedVector{$elty})
+        function tfttr!(A::StridedMatrix{$elty}, transr::Char, uplo::Char, Arf::StridedVecOrMat{$elty})
```
Comment (andreasnoack) [inline src/lapack.jl]:
```suggestion
        function tfttr!(A::StridedMatrix{$elty}, transr::Char, uplo::Char, Arf::VecOrMat{$elty})
```

File: src/lapack.jl [hunk 10]
```diff
@@ -336,13 +344,15 @@ for (f, elty) in (
 )
 
     @eval begin
-        function trttf!(transr::Char, uplo::Char, A::StridedMatrix{$elty})
+        function trttf!(Arf::StridedVecOrMat{$elty}, transr::Char, uplo::Char, A::StridedMatrix{$elty})
```
Comment (andreasnoack) [inline src/lapack.jl]:
```suggestion
        function trttf!(Arf::VecOrMat{$elty}, transr::Char, uplo::Char, A::StridedMatrix{$elty})
```

File: src/triangular.jl [hunk 11]
```diff
@@ -0,0 +1,57 @@
+struct TriangularRFP{T<:BlasFloat} <: AbstractRFP{T}
+    data::Matrix{T}
+    transr::Char
+    uplo::Char
+end
+
+function TriangularRFP(A::Matrix{T}, uplo::Symbol = :U; transr::Symbol=:N) where {T}
+    n = checksquare(A)
+    (ul = first(string(uplo))) ∈ "UL" ||
+        throw(ArgumentError("uplo = $uplo should be :U or :L"))
```
Comment (andreasnoack) [inline src/triangular.jl]:
```suggestion
    ul = first(string(uplo))
    ul ∉ "UL"
        throw(ArgumentError("uplo = $uplo should be :U or :L"))
    end
```

File: src/triangular.jl [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    (tr = first(string(transr))) ∈ (T <: Complex ? "NC" : "NT") ||
+        throw(ArgumentError("transr = $transr should be :N or :(T <: Complex ? :C : :T)"))
```
Comment (andreasnoack) [inline src/triangular.jl]:
```suggestion
    tr = first(string(transr))
    tr ∉ (T <: Complex ? "NC" : "NT")
        throw(ArgumentError("transr = $transr should be :N or :(T <: Complex ? :C : :T)"))
    end
```

File: src/utilities.jl [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+Returns the size of the `data` field in an RFP array with `transr == 'N'`, representing a matrix of size `n`
+"""
+function _parentsize(n::Integer, tr::Bool = false)
+    n > 0 || throw(ArgumentError("n = $n must be positive"))
+    sz = (n + iseven(n), (n + 1) >> 1)
+    return tr ? reverse(sz) : sz
+end
+
+Base.parent(A::AbstractRFP) = A.data
```
Comment (andreasnoack) [inline src/utilities.jl]: Let's also drop this one

**Review by andreasnoack — COMMENTED** [review]
Looks good. A few minor suggestions. Codecov also points out a few code paths that aren't covered by tests. It would be good to have all the coed paths covered.

File: src/utilities.jl (hunk 13 above)
Comment (dmbates) [reply src/utilities.jl]: There were a couple of 'if's missed in the editing of `src/triangular.jl` and it took me an embarrassingly long time to decide why compilation was failing.  (The error message occurred later.)

**Comment by dmbates** [comment]
@andreasnoack Let me know if you think this is ready to go.  @ararslan It is probably not necessary for you to do a line-by-line review - I am mostly interested in glitches in style, etc.

**Comment by andreasnoack** [comment]
It looks like there are two unaddressed suggestions in the lapack.jl file

**Comment by dmbates** [comment]
Thanks for pointing out the omissions.  Those changes are now incorporated.

**Review by andreasnoack — APPROVED** [review]

**Pull request merged**